rampwf/hyperopt/hyperopt.py

```diff
@@ -0,0 +1,521 @@
+        # workflow element names, values are lists are hypers belonging
+        # to the workflow element
+        self.hypers_per_workflow_element = {
+            wen: [] for wen in self.problem.workflow.element_names}
+        for h in self.hyperparameters:
+            self.hypers_per_workflow_element[h.workflow_element_name].append(h)
+
+        # Set up df_scores_ which will contain one row per experiment
+        scores_columns = ['fold_i']
+        scores_columns += self.hyperparameter_names
+        scores_columns += ['train_' + name for name in self.score_names]
+        scores_columns += ['valid_' + name for name in self.score_names]
+        scores_columns += ['train_time', 'valid_time', 'n_train', 'n_valid']
+        dtypes = ['int'] + [h.dtype for h in self.hyperparameters] +\
+            ['float'] * 2 * len(self.score_names) + ['float'] * 2 + ['int'] * 2
+        self.df_scores_ = pd.DataFrame(columns=scores_columns)
+        for column, dtype in zip(scores_columns, dtypes):
+            self.df_scores_[column] = self.df_scores_[column].astype(dtype)
+
+    def _run_next_experiment(self, module_path, fold_i):
+        _, _, df_scores = run_submission_on_cv_fold(
+            self.problem, module_path=module_path, fold=self.cv[fold_i],
+            X_train=self.X_train, y_train=self.y_train)
+        return df_scores
+
+    def _update_df_scores(self, df_scores, fold_i):
+        row = {'fold_i': fold_i}
+        for h in self.hyperparameters:
+            row[h.name] = h.default
+        for name in self.score_names:
+            row['train_' + name] = df_scores.loc['train'][name]
+            row['valid_' + name] = df_scores.loc['valid'][name]
+        row['train_time'] = float(df_scores.loc['train']['time'])
+        row['valid_time'] = float(df_scores.loc['valid']['time'])
+        row['n_train'] = len(self.cv[fold_i][0])
+        row['n_valid'] = len(self.cv[fold_i][1])
+        self.df_scores_ = self.df_scores_.append(row, ignore_index=True)
+
+    def _make_and_save_summary(self, hyperopt_output_path):
+        summary_groupby = self.df_scores_.groupby(
+            self.hyperparameter_names)
+        means = summary_groupby.mean().drop(columns=['fold_i'])
+        stds = summary_groupby.std().drop(columns=['fold_i'])
+        counts = summary_groupby.count()[['n_train']].rename(
+            columns={'n_train': 'n_folds'})
+        self.df_summary_ = pd.merge(
+            means, stds, left_index=True, right_index=True,
+            suffixes=('_m', '_s'))
+        self.df_summary_ = pd.merge(
+            counts, self.df_summary_, left_index=True, right_index=True)
+        print(self.df_summary_)
+        summary_fname = os.path.join(hyperopt_output_path, 'summary.csv')
+        self.df_summary_.to_csv(summary_fname)
+
+    def _save_best_model(self):
+        official_scores = self.df_summary_[
+            'valid_' + self.problem.score_types[0].name + '_m']
+        best_defaults = official_scores.idxmax()
```

Comment:
depending on `is_lower_the_better` value, this can be `idxmax()` or `idxmin()`

Reply:
done, nice catch!